src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs

```diff
@@ -742,6 +705,112 @@ private async Task<DownloadedBuild> GatherDropForBuildAsync(Build build, string
             return downloadedAsset;
         }
 
+        /// <summary>
+        /// Download a single asset from any of its locations. Iterate over the asset's locations
+        /// until the asset download succeed or all locations have been tested.
+        /// </summary>
+        private async Task<DownloadedAsset> DownloadAssetFromAnyLocationAsync(HttpClient client,
+                                                                            Build build, 
+                                                                            Asset asset, 
+                                                                            List<AssetLocation> assetLocations, 
+                                                                            string rootOutputDirectory, 
+                                                                            List<string> errors)
+        {
+            // Walk the locations and attempt to gather the asset at each one, setting the output
+            // path based on the type. Stops at the first successfull download.
+            foreach (AssetLocation location in assetLocations)
+            {
+                var downloadedAsset = await DownloadAssetFromLocation(client,
+                    build,
+                    asset,
+                    location,
+                    rootOutputDirectory,
+                    errors);
+
+                if (downloadedAsset.Successful)
+                {
+                    return downloadedAsset;
+                }
+            }
+
+            return new DownloadedAsset()
+            {
+                Successful = false,
+                Asset = asset
+            };
+        }
+
+        /// <summary>
+        /// Download a single asset from the latest asset location registered.
+        /// </summary>
+        private async Task<DownloadedAsset> DownloadAssetFromLatestLocation(HttpClient client,
+                                                                            Build build,
+                                                                            Asset asset,
+                                                                            List<AssetLocation> assetLocations,
+                                                                            string rootOutputDirectory,
+                                                                            List<string> errors)
+        {
+            AssetLocation latestLocation = assetLocations.OrderBy(al => al.Id).First();
```

Comment:
Tiny nit: kinda odd scenario but we could check if `assetLocations` is empty to make sure calling First() won't throw. If we are certain that this will always have items in it, then this is a no op.